node-jose/index.d.ts

```diff
@@ -0,0 +1,1380 @@
+     *
+     * Handlers are intended to support 'crit' extensions. When a boolean value,
+     * the member is expected to be processed once decryption is fully complete.
+     * When a function, it is called just before the ciphertext is decrypted
+     * (processed as if it were a `prepare` handler, as decribed in `decryptCritProcessors`).
+     * Note that normal processing of `decrypt()` does not continue until all
+     * relevant handlers have completed. Any changes handlers make to the
+     * provided objects affects `decrypt()`'s processing.
+     */
+    type decryptCritHeaderOptions = {
+        handlers: { [crit_header: string]: boolean | Function | decryptCritProcessors }
+    };
+
+    /**
+     * For the critical header processing, the options may include an object like this,
+     * which defines these fields to run at various times of decryption processing.
+     * @property {Function} recipient A function called after a valid key is
+     *           determined; it takes an object describing the recipient, and
+     *           returns a Promise that is fulfilled once the handler's processing
+     *           is complete.
+     * @property {Function} prepare A function called just prior to decrypting
+     *           the ciphertext; it takes an object describing the decryption
+     *           result (but containing `ciphertext` and `tag' instead of `payload`
+     *           and `plaintext`), and returns a Promise that is fulfilled once
+     *           the handler's processing is complete.
+     * @property {Function} complete A function called once decryption is complete,
+     *           just prior to fulfilling the Promise returned by `decrypt()`; it
+     *           takes the object that will be returned by `decrypt()`'s fulfilled
+     *           Promise, and returns a Promise that is fulfilled once the handler's
+     *           processing is complete.
+     */
+    type decryptCritProcessors = {
+        recipient?: Function,
+        prepare?: Function,
+        complete?: Function
+    };
+
+    /**
+     * Object that defines the output that is given from the `decrypt()` function
+     * of a `Decryptor`.
+     *
+     * @property {any} header The JOSE Header, combined from the relevant "header" and
+     *           "protected" fields from the original JWE object.
+     * @property {string[]} protected An array containing the names of the protected fields
+     * @property {JWK.Key} key The used to decrypt the content
+     * @property {Buffer} payload The decrypted content (as a Buffer)
+     * @property {Buffer} plaintext An alias for `payload`
+     */
+    type decryptReturn = {
+        header: any,
+        protected: string[],
+        key: JWK.Key,
+        payload: Buffer,
+        plaintext: Buffer
+    };
+
+    /**
+     * The options for {@link createEncrypt}.
+     *
+     * @property {boolean|string} zip Determines the compression algorithm to
```

Comment:
The `string`s for this type should be broken into their allowed values, eg `zip: boolean | "DEF" | ""`.